Composer/packages/client/src/pages/language-understanding/code-editor.tsx

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+/* eslint-disable react/display-name */
+import React, { useState, useEffect, useMemo, useContext, useCallback } from 'react';
+import { LuEditor } from '@bfc/code-editor';
+import get from 'lodash/get';
+import debounce from 'lodash/debounce';
+import isEmpty from 'lodash/isEmpty';
+import { editor } from '@bfcomposer/monaco-editor/esm/vs/editor/editor.api';
+import { luIndexer, combineMessage, isValid, filterTemplateDiagnostics } from '@bfc/indexers';
+import { RouteComponentProps } from '@reach/router';
+import querystring from 'query-string';
+
+import { StoreContext } from '../../store';
+import * as luUtil from '../../utils/luUtil';
+
+const { parse } = luIndexer;
+
+const lspServerPath = '/lu-language-server';
+
+interface CodeEditorProps extends RouteComponentProps<{}> {
+  fileId: string;
+}
+
+const CodeEditor: React.FC<CodeEditorProps> = props => {
+  const { actions, state } = useContext(StoreContext);
+  const { luFiles } = state;
+  const { fileId } = props;
+  const file = luFiles?.find(({ id }) => id === fileId);
+  const [diagnostics, setDiagnostics] = useState(get(file, 'diagnostics', []));
+  const [errorMsg, setErrorMsg] = useState('');
+  const [luEditor, setLuEditor] = useState<editor.IStandaloneCodeEditor | null>(null);
+
+  const search = props.location?.search ?? '';
+  const searchSectionName = querystring.parse(search).t;
+  const sectionId = Array.isArray(searchSectionName)
+    ? searchSectionName[0]
+    : typeof searchSectionName === 'string'
+    ? searchSectionName
+    : undefined;
+  const intent = sectionId && file ? file.intents.find(({ Name }) => Name === sectionId) : undefined;
+
+  const hash = props.location?.hash ?? '';
+  const hashLine = querystring.parse(hash).L;
+  const line = Array.isArray(hashLine) ? +hashLine[0] : typeof hashLine === 'string' ? +hashLine : undefined;
+
+  const inlineMode = !!intent;
+  const [content, setContent] = useState(intent?.Body || file?.content);
+
+  useEffect(() => {
+    // reset content with file.content initial state
+    if (!file || isEmpty(file) || content) return;
+    const value = intent ? intent.Body : file.content;
+    setContent(value);
+  }, [file, sectionId]);
+
+  useEffect(() => {
+    const currentDiagnostics = inlineMode && intent ? filterTemplateDiagnostics(diagnostics, intent) : diagnostics;
+    const isInvalid = !isValid(currentDiagnostics);
+    const text = isInvalid ? combineMessage(diagnostics) : '';
+    setErrorMsg(text);
+  }, [diagnostics]);
```

Comment:
About the useEffect hook here.

`diagnostics` comes from local state and is always changed by `setDiagnostics`;
`errorMsg` always changes along with `diagnostic` and there is no asyn transformation.

If so, `errorMsg` could be treated as a computed value out of `diagnostics`. Use `useMemo` hook instead of `useEffect` will have better perf.

Reply:
The displayed `'errorMsg'` consists of two parts, one computed from `diagnostics`, another yielded from catched error. They can be managed separately and joined together as the final value.
```
errorMsg={computedFromDiagnostics || state.errorMsg}
```

Reply:
Updated, great review!